omx/gstomxcamera.c

```diff
@@ -0,0 +1,1214 @@
+        "Failed to set memory type on port %lu: %s (0x%08x)",
+        mem_type.nPortIndex, gst_omx_error_to_string (err), err);
+    return FALSE;
+  }
+  GST_DEBUG_OBJECT (self, "Memory type: %d", mem_type.eBufMemoryType);
+
+  GST_OMX_INIT_STRUCT (&hw_port);
+  /* Set capture interface */
+  hw_port.eHwPortId = self->interface;
+  err = gst_omx_component_set_parameter (self->comp,
+      OMX_TI_IndexParamVFCCHwPortID, &hw_port);
+  GST_DEBUG_OBJECT (self, "Hardware port id: %d", hw_port.eHwPortId);
+
+  GST_OMX_INIT_STRUCT (&hw_port_param);
+  hw_port_param.eCaptMode = self->capt_mode;
+  hw_port_param.eVifMode = self->vip_mode;
+  hw_port_param.eInColorFormat = OMX_COLOR_FormatYCbYCr;
+  hw_port_param.eScanType = self->scan_type;
+  hw_port_param.nMaxHeight = GST_VIDEO_INFO_HEIGHT (info);
+  hw_port_param.nMaxWidth = GST_VIDEO_INFO_WIDTH (info);
+  hw_port_param.nMaxChnlsPerHwPort = 1;
+  if (self->scan_type == OMX_VIDEO_CaptureScanTypeInterlaced)
+    hw_port_param.nMaxHeight = hw_port_param.nMaxHeight >> 1;
+
+  err = gst_omx_component_set_parameter (self->comp,
+      OMX_TI_IndexParamVFCCHwPortProperties, &hw_port_param);
+
+  GST_DEBUG_OBJECT (self,
+      "Hw port properties: capture mode %d, vif mode %d, max height %li, max width %li, max channel %li, scan type %d, format %d",
+      hw_port_param.eCaptMode, hw_port_param.eVifMode, hw_port_param.nMaxHeight,
+      hw_port_param.nMaxWidth, hw_port_param.nMaxChnlsPerHwPort,
+      hw_port_param.eScanType, hw_port_param.eInColorFormat);
+
+  gst_omx_camera_set_skip_frames (self);
+
+  /* Configure output pool */
+  config = gst_buffer_pool_get_config (self->outpool);
+  gst_buffer_pool_config_set_params (config, caps,
+      port_def.nBufferSize, port_def.nBufferCountActual,
+      port_def.nBufferCountActual);
+  if (!gst_buffer_pool_set_config (self->outpool, config))
+    goto config_pool_failed;
+
+  /* Calculate duration based on the framerate to use it in case
+   * nTickCount is not provided by the OMX component */
+  self->duration = gst_util_uint64_scale_int (GST_SECOND,
+      GST_VIDEO_INFO_FPS_D (info), GST_VIDEO_INFO_FPS_N (info));
+
+  return TRUE;
+
+drain_failed:
+  {
+    GST_ERROR_OBJECT (self, "Failed to reconfigure, cannot drain component");
+    return FALSE;
+  }
+config_pool_failed:
+  {
+    GST_INFO_OBJECT (self, "Failed to set config on output pool");
+    return FALSE;
+  }
+}
+
+static gboolean
+gst_omx_camera_set_caps (GstBaseSrc * src, GstCaps * caps)
+{
+  GstOMXCamera *self = GST_OMX_CAMERA (src);
+  GstVideoInfo info;
+  gchar *caps_str = NULL;
+
+  GST_INFO_OBJECT (self, "set caps (%" GST_PTR_FORMAT "): %s", caps, caps_str =
+      gst_caps_to_string (caps));
+  if (caps_str)
+    g_free (caps_str);
+
+  gst_video_info_init (&info);
+  if (G_UNLIKELY (!gst_video_info_from_caps (&info, caps)))
+    return FALSE;
+
+  if (!gst_omx_camera_set_format (self, caps, &info))
+    return FALSE;
+
+  return TRUE;
+}
+
+
+static gboolean
```

Comment:
Not necessary, don't implement.

Reply:
As change_state function is avoided 4e962806a3de137cdc02851f967289f6ef2021b1 _start call is not empty any more.